tiproxy: add tiproxy (#15634)
pingcap/docs: tiproxy/tiproxy-configuration.md

---
title: Configure TiProxy
summary: Learn how to configure TiProxy.
aliases: ['/docs/dev/tiproxy/tiproxy-configuration/','/docs/dev/reference/tiproxy/configuration/']
---

# Configure TiProxy

This document introduces the configuration parameters related to the deployment and use of TiProxy.

## Configure the `tiproxy.toml` file

This section introduces the configuration parameters of TiProxy.

> **Tip:**
>
> If you need to adjust the value of a configuration item, refer to [Modify the configuration](/maintain-tidb-using-tiup.md#modify-the-configuration).

> ```toml

Author: 改掉了, 再看看？

---
title: Configure TiProxy
summary: Learn how to configure TiProxy.
---

# Configure TiProxy

This document introduces the configuration parameters related to the deployment and use of TiProxy. The following is an example configuration:

## Configure the `tiproxy.toml` file

This section introduces the configuration parameters of TiProxy.

> **Tip:**
>
> If you need to adjust the value of a configuration item, refer to [Modify the configuration](/maintain-tidb-using-tiup.md#modify-the-configuration). Since TiProxy will do config hot-reloading, you can also skip restart by `tiup cluster reload --skip-restart`.

### proxy

Configuration for SQL port.

#### `addr`

+ Default Value: `0.0.0.0:6000`
+ SQL gateway address. You should specify `ip:port`.

#### `graceful-wait-before-shutdown`

+ Default Value: `0`

> + Hot-reload supported.

Author: 对的

+ HTTP status returns unhealthy and the SQL port accepts new connections for the last `graceful-wait-before-shutdown` seconds. After that, refuse new connections and drain clients. It is recommanded to be set to 0 when there's no other proxy(e.g. NLB) between the client and TiProxy.

#### `graceful-close-conn-timeout`

+ Default Value: `15`
+ Hot-reload supported.
+ Close connections when they have finished current transactions (AKA drain clients) after `graceful-close-conn-timeout` seconds. It is recommanded to be set longer than the lifecycle of a transaction.

#### `max-connections`

+ Default Value: `0`
+ Hot-reload supported.
+ Accept as many as `max-connections` connections. Zero means no limitation.

#### `conn-buffer-size`

> + Default Value: `0`

Author: 代表自动管理, 补充描述了

+ Hot-reload supported.
+ Tradeoff between memory and performance. Larger buffer may yield better performance result.

#### `pd-addrs`

+ Default Value: `127.0.0.1:2379`

> + Automatically discovery TiDB instances and set them as backend.

Author: 对

#### `proxy-protocol`

+ Default Value: ``
+ Hot-reload supported.

> + Enable proxy protocol handling on the port. You could specify `v2` to handle proxy protocol version 2.

Author: 对. proxy protocol有两个版本

+ Require TLS on backend instances.

### api

Configurations for HTTP gateway.

#### `addr`

+ Default Value: `0.0.0.0:3090`
+ API gateway address. You should specify `ip:port`.

#### `proxy-protocol`

+ Default Value: ``
+ Enable proxy protocol handling on the port. You could specify `v2` to handle proxy protocol version 2.

### log

#### `level`

+ Default Value: `info`
+ Hot-reload supported.
+ You can specify:

    + `tidb`: formats used by tidb, check https://github.com/tikv/rfcs/blob/master/text/0018-unified-log-format.md
    + `json`: structured json formats
    + `console`: log format for human.

### log.log-file

#### `filename`

+ Default Value: ``
+ Hot-reload supported.
+ Log file path. Non empty value will enable logging to file.

#### `max-size`

+ Default Value: `300`
+ Hot-reload supported.
+ Log file maximum size, in megabytes. Log will be rotated.

#### `max-days`

+ Default Value: `3`
+ Hot-reload supported.
+ Maximum days to retain old log files. It is deleted once it is too old.

#### `max-backups`

+ Default Value: `3`
+ Hot-reload supported.
+ Maximum number of log files. Extra log files will be deleted once there are too many.

### security

> There are 4 tls objects in `[security]` section. They share same configuration formats and fields. But they are interpreted differently according to their usage.

Author: 是指有`xxx-tls`一共四个小节, 分为两个类型, 而且四小节所有可以指定的field都一样.

---
title: Configure TiProxy
summary: Learn how to configure TiProxy.
---

# Configure TiProxy

This document introduces the configuration parameters related to the deployment and use of TiProxy. The following is an example configuration:

> ```

Reviewer: ```suggestion
```toml
```

> ```toml

Reviewer 2: yaml?

[proxy]
addr = "0.0.0.0:6000"
max-connections = 100

[api]
addr = "0.0.0.0:3080"

[log]
level = "info"

[security]
	[security.cluster-tls]
	skip-ca = true

	[security.sql-tls]

> 	skip-ca = true

Reviewer 2: indent?

+ Support hot-reload: yes
+ You can specify:

    + `tidb`: format used by TiDB. For details, refer to [Unified Log Format](https://github.com/tikv/rfcs/blob/master/text/0018-unified-log-format.md).
    + `json`: structured JSON format.
    + `console`: human-readable log format.

### log.log-file

#### `filename`

+ Default Value: ``
+ Support hot-reload: yes
+ Log file path. Non empty value will enable logging to file.

#### `max-size`

+ Default Value: `300`
+ Support hot-reload: yes
+ Specifies the maximum size, in megabytes, for log files. Logs will be rotated.

#### `max-days`

+ Default Value: `3`
+ Support hot-reload: yes
+ Specifies the maximum number of days to keep old log files. Outdated log files are deleted after surpassing this period.

#### `max-backups`

+ Default Value: `3`
+ Support hot-reload: yes
+ Specifies the maximum number of log files to be retained. Surplus log files will be automatically deleted when an excessive number is reached.

### security

There are four TLS objects in the `[security]` section with different names. They share the same configuration format and fields, but they are interpreted differently depending on their names.

> ```

Reviewer: ```suggestion
```toml
```

---
title: Configure TiProxy
summary: Learn how to configure TiProxy.
---

# Configure TiProxy

This document introduces the configuration parameters related to the deployment and use of TiProxy. The following is an example configuration:

```toml
[proxy]
addr = "0.0.0.0:6000"
max-connections = 100

[api]
addr = "0.0.0.0:3080"

[log]
level = "info"

[security]
[security.cluster-tls]
skip-ca = true

[security.sql-tls]
skip-ca = true
```

> # ignore-wrong-namespace = true

Reviewer 3: 这里是标题吗？还是写错了

Author: 错了.

#### `require-backend-tls`

+ Default value: `true`
+ Support hot-reload: yes
+ Require TLS between TiProxy and TiDB servers. If the TiDB server doesn't support TLS, clients will report an error when connecting to TiProxy.

### api

Configurations for HTTP gateway.

#### `addr`

+ Default value: `0.0.0.0:3090`
+ API gateway address. You can specify `ip:port`.

#### `proxy-protocol`

+ Default value: ``
+ Possible values: `v2`
+ Enable proxy protocol handling on the port. You can specify `v2` to handle proxy protocol version 2. `v1` is not supported.

### log

#### `level`

+ Default value: `info`
+ Support hot-reload: yes
+ Possible values: `debug`, `info`, `warn`, `error`, `panic`
+ Specify log level. With `panic` level, TiProxy will panic on errors.

#### `encoder`

+ Default value: `tidb`
+ You can specify:

    + `tidb`: format used by TiDB. For details, refer to [Unified Log Format](https://github.com/tikv/rfcs/blob/master/text/0018-unified-log-format.md).
    + `json`: structured JSON format.
    + `console`: human-readable log format.

### log.log-file

#### `filename`

+ Default value: ``
+ Support hot-reload: yes

> + Log file path. Non empty value will enable logging to file.

Reviewer 2: ```suggestion
+ Log file path. Non empty value will enable logging to file. When TiProxy is deployed with TiUP, the filename is set automatically.
```

#### `max-size`

+ Default value: `300`
+ Support hot-reload: yes

> + Specifies the maximum size, in megabytes, for log files. Logs will be rotated.

Reviewer 2: ```suggestion
+ Unit: MB
+ Specifies the maximum size for log files. A log file will be rotated if its size exceeds this limit.
```

#### `max-days`

+ Default value: `3`
+ Support hot-reload: yes
+ Specifies the maximum number of days to keep old log files. Outdated log files are deleted after surpassing this period.

#### `max-backups`

+ Default value: `3`
+ Support hot-reload: yes
+ Specifies the maximum number of log files to be retained. Surplus log files will be automatically deleted when an excessive number is reached.

### security

There are four TLS objects in the `[security]` section with different names. They share the same configuration format and fields, but they are interpreted differently depending on their names.

```toml
[security]
    [sql-tls]
    skip-ca = true
    [server-tls]
    auto-certs = true
```

All TLS options are hot-reloaded.

TLS object fields:

+ `ca`: specifies the CA
+ `cert`: specifies the certificate
+ `key`: specifies the private key
+ `auto-certs`: mostly used for tests. It generates certificates if no certificate/key is specified.
+ `skip-ca`: skips verifying certificates using CA on client object or skips server-side verification on server object.

> + `min-tls-version`: sets the minimum TLS version.

Reviewer 2: ```suggestion
+ `min-tls-version`: sets the minimum TLS version. Possible values are `1.0`、`1.1`、`1.2`, and `1.3`. The default value is `1.1`, which allows v1.1 or higher TLS versions.
```

+ `rsa-key-size`: sets the RSA key size when `auto-certs` is enabled.
+ `autocert-expire-duration`: sets the default expiration duration for auto-generated certificates.

Objects are classified into client or server objects by their names.

For client TLS object:

- You must set either `ca` or `skip-ca` to skip verifying server certificates.
- Optionally, you can set `cert`/`key` to pass server-side client verification.
- Useless fields: auto-certs.

For server TLS object:

> + You must set either `cert`/`key` or `auto-certs` to generate a temporary certificate, mainly for testing purposes.

Reviewer 2: ```suggestion
+ You can set either `cert`/`key` or `auto-certs` to support TLS connections. Otherwise, TiProxy doesn't support TLS connections.
```

#### `max-connections`

+ Default value: `0`
+ Support hot-reload: yes
+ Each TiProxy instance can accept `max-connections` connections at most. `0` means no limitation.

#### `conn-buffer-size`

+ Default value: `32768`
+ Support hot-reload: yes, but only for new connections
+ Range: `[1024, 16777216]`
+ This configuration item lets you decide the connection buffer size. Each connection uses one read buffer and one write buffer. It is a tradeoff between memory and performance. A larger buffer might yield better performance results but consume more memory. When it is `0`, TiProxy uses the default buffer size.

#### `pd-addrs`

+ Default value: `127.0.0.1:2379`
+ Support hot-reload: no
+ The PD addresses TiProxy connects to. TiProxy discovers TiDB instances by fetching the TiDB list from the PD. It is set automatically when TiProxy is deployed by TiUP or TiDB Operator.

#### `proxy-protocol`

+ Default value: ``
+ Support hot-reload: yes, but only for new connections
+ Possible values: ``, `v2`

> + Enable the [PROXY protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) on the port. By enabling the PROXY protocol, TiProxy can pass the real client IP address to TiDB. `v2` indicates using the PROXY protocol version 2, and `` indicates disabling the PROXY protocol. If the PROXY protocol is enabled on TiProxy, you need to also enable the [PROXY protocol](/tidb-configuration-file/#proxy-protocol) on the TiDB server.

Reviewer 3: ```suggestion
+ Enable the [PROXY protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) on the port. By enabling the PROXY protocol, TiProxy can pass the real client IP address to TiDB. `v2` indicates using the PROXY protocol version 2, and `` indicates disabling the PROXY protocol. If the PROXY protocol is enabled on TiProxy, you need to also enable the [PROXY protocol](/tidb-configuration-file.md#proxy-protocol) on the TiDB server.
```

#### `require-backend-tls`

+ Default value: `true`
+ Support hot-reload: yes, but only for new connections
+ Require TLS between TiProxy and TiDB servers. If the TiDB server doesn't support TLS, clients will report an error when connecting to TiProxy.

### api

Configurations for HTTP gateway.

#### `addr`

+ Default value: `0.0.0.0:3090`
+ Support hot-reload: no
+ API gateway address. You can specify `ip:port`.

#### `proxy-protocol`

+ Default value: ``
+ Support hot-reload: no
+ Possible values: ``, `v2`
+ Enable the [PROXY protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) on the port. `v2` indicates using the PROXY protocol version 2, and `` indicates disabling the PROXY protocol.

### log

#### `level`

+ Default value: `info`
+ Support hot-reload: yes
+ Possible values: `debug`, `info`, `warn`, `error`, `panic`

> + Specify log level. With `panic` level, TiProxy will panic on errors.

Reviewer 3: ```suggestion
+ Specify the log level. With the `panic` level, TiProxy will panic on errors.
```
